### PR TITLE
[Go] Interval

### PR DESCRIPTION
**Pull request by neatnerd**
Added additional interval math functionality

### CODE REVIEW

**Comment by neatnerd** [comment]
Yo @davesisson I will write the test tonight then, but only for Go target and only this section, ok?

**Comment by davesisson** [comment]
Thanks!  That would be great!

**Comment by neatnerd** [comment]
There is one issue that I have stumble across during testing
Why does Go target actually not include Stop position of token? as mention here:

```go
/* stop is not included! */
func NewInterval(start, stop int) *Interval {
	i := new(Interval)
	i.Start = start
	i.Stop = stop
	return i
}
```

Actually the calculation of length then is also adjusted accordingly:
Go:
```go
func (i *Interval) length() int {
	return i.Stop - i.Start
}
```
vs
Java:
```java
/** return number of elements between a and b inclusively. x..x is length 1.
 *  if b &lt; a, then length is 0.  9..10 has length 2.
 */
public int length() {
	if ( b<a ) return 0;
	return b-a+1;
}
```
vs
C++:
```cpp
size_t Interval::length() const {
  if (b < a) {
    return 0;
  }
  return size_t(b - a + 1);
}
```
I did the blame and seems it has not changed since initial commit. @pboyer @davesisson @parrt do you, guys, know why?

Reason I am asking is cause it is a fundamental difference in logic in many parts of application imho

**Comment by davesisson** [comment]
I'm not sure why it is this way -- perhaps it was thought that it was simpler.

But I guess it really comes down to -- are you really going to normalize the behavior at this point?  I would expect that changing it to match (if that happens) would happen as a separate PR.

**Comment by neatnerd** [comment]
Ok, that means, I have to adjust logic for Interval position that put into this PR, ok?

**Comment by parrt** [comment]
Sounds like a bug in Go target. length() makes no sense as 1 less than the length I guess.

**Comment by neatnerd** [comment]
@parrt I dont think it is a bug, it just uses half-opened intervals instead of closed interval, thus calculation is ok. There are in general two options: adjust this logic or adjust everything around it

**Comment by parrt** [comment]
Oh right. yeah, i like a..b inclusive but [a,b) is more typical. hmm... as long as the doc/API is clear which way it goes, maybe we leave for compatibility?

**Comment by neatnerd** [comment]
To be honest, this part is pretty undocumented, the only reference I found in official docs is [here](http://www.antlr.org/api/Java/index.html) with 

> public int length()
return number of elements between a and b inclusively. x..x is length 1. if b < a, then length is 0. 9..10 has length 2.

So I would say the call how we want to keep it is up to you [a,b] or [a, b)
Also maybe it would be beneficial to add it to Interval class docu

**Comment by neatnerd** [comment]
@parrt any thoughts on this maybe?

**Comment by parrt** [comment]
To be consistent with other targets I suggest using a..b inclusive `[a,b]`

**Comment by parrt** [comment]
@davesisson there are tests for interval: TestIntervalSet.java :)
Ter

**Comment by pboyer** [comment]
I'm liking this change! :D

File: runtime/Go/antlr/interval_set.go [hunk 1]
```diff
@@ -27,6 +27,68 @@ func (i *Interval) Contains(item int) bool {
...
+
+// Return the interval in common between this and other
+func (i *Interval) Intersection(other *Interval) *Interval{
+	return NewInterval(intMax(i.Start, other.Start), intMin(i.Stop, other.Stop))
+}
+//  Return the interval with elements from this not in other;
+//  other must not be totally enclosed (properly contained)
+//  within this, which would result in two disjoint intervals
+//  instead of the single one returned by this method.
+func (i *Interval) DifferenceNotProperlyContained(other *Interval) *Interval{
```
Comment (pboyer) [inline runtime/Go/antlr/interval_set.go]: It's Go style to prefix any method comment with the name of the function. So, this should read `DifferenceNotProperlyContained returns the interval...`

**Comment by neatnerd** [comment]
@parrt Regarding tests, yes there are tests for IntervalSet, but there are no tests for Interval.

**Comment by parrt** [comment]
@neatnerd ah! ok.

**Comment by parrt** [comment]
Unfortunately I assumed you had tested this. It breaks all of the unit tests so I'm going to have to revert.

**Comment by neatnerd** [comment]
Yeah, sorry, wanted to fix it this weekend. Shall I make a new PR?

**Comment by parrt** [comment]
yes, please. seems like it'll change many many locations in the API, right?

**Comment by neatnerd** [comment]
Well... There were couple of inconsistencies as to how intervals are handled. For example, I have assumed that if I delete from interval [10;20], point 15, result shall be [10;14] and [16;20]. Seems like it has to be [10;15] and [16;20]. Is there a good set of tests/documentation somewhere what I can rely upon for intervals? I assume no, right?

**Comment by parrt** [comment]
nope. We made the assumption that all intervals were inclusive so anything that deviates from that is likely to be wrong or inconsistent.

**Comment by neatnerd** [comment]
Ok, thx. Will research it a bit more in detail and make a separate PR

**Pull request merged**